docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst

```diff
@@ -0,0 +1,356 @@
+.. _analyzing-data-flow-in-python:
+
+Analyzing data flow in Python
+=============================
+
+You can use CodeQL to track the flow of data through a Python program to places where the data is used.
+
+About this article
+------------------
+
+This article describes how data flow analysis is implemented in the CodeQL libraries for Python and includes examples to help you write your own data flow queries.
+The following sections describe how to use the libraries for local data flow, global data flow, and taint tracking.
+For a more general introduction to modeling data flow, see ":ref:`About data flow analysis <about-data-flow-analysis>`."
+
+Local data flow
+---------------
+
+Local data flow is data flow within a single method or callable. Local data flow is easier, faster, and more precise than global data flow, and is sufficient for many queries.
+
+Using local data flow
+~~~~~~~~~~~~~~~~~~~~~
+
+The local data flow library is in the module ``DataFlow``, which defines the class ``Node`` denoting any element that data can flow through. The ``Node`` class has a number of useful subclasses, such as ``ExprNode`` for expressions, ``CallCfgNode`` for function and method calls, and ``ParameterNode`` for parameters, all of which are subclasses of ``CfgNode`` which holds all those data-flow nodes which are associated with a control-flow node. You can map between data flow nodes and expressions/control-flow nodes using the member predicates ``asExpr`` and ``asCfgNode``:
```

Comment:
> all of which are subclasses of ``CfgNode`` which holds all those data-flow nodes which are associated with a control-flow node.

I'm not sure what the point you're trying to make with this text is. 

I think splitting it up to be it's own sentence, or moving it like in the suggestion could improve things a bit. But really, I'm unsure, since I don't quite understand the intent :sweat_smile: 

```suggestion
The local data flow library is in the module ``DataFlow``, which defines the class ``Node`` denoting any element that data can flow through. The ``Node`` class has a number of useful subclasses, such as ``ExprNode`` for expressions, ``CfgNode`` for control-flow nodes, ``CallCfgNode`` for function and method calls, and ``ParameterNode`` for parameters. You can map between data flow nodes and expressions/control-flow nodes using the member predicates ``asExpr`` and ``asCfgNode``:
```

Reply:
The intent was to also express the type hierarchy, but I agree that it is trying to do too much. Subtly hinting at it, as in your suggestion, works much better.